packages/trader/src/AppV2/Components/TradeParameters/RiskManagement/take-profit-and-stop-loss-input.tsx

```diff
@@ -0,0 +1,328 @@
+    const is_be_response_received_ref = parent_is_be_response_received_ref || is_be_response_received;
+
+    const [is_enabled, setIsEnabled] = React.useState(is_take_profit_input ? has_take_profit : has_stop_loss);
+    const [new_input_value, setNewInputValue] = React.useState(is_take_profit_input ? take_profit : stop_loss);
+    const [error_text, setErrorText] = React.useState(initial_error_text ?? '');
+
+    // Refs for handling focusing and bluring input
+    const input_ref = React.useRef<HTMLInputElement>(null);
+    const focused_input_ref = React.useRef<HTMLInputElement>(null);
+    const focus_timeout = React.useRef<ReturnType<typeof setTimeout>>();
+
+    const contract_types = getDisplayedContractTypes(trade_types, contract_type, trade_type_tab);
+    const decimals = getDecimalPlaces(currency);
+    const currency_display_code = getCurrencyDisplayCode(currency);
+    const Component = should_wrap_with_actionsheet ? ActionSheet.Content : 'div';
+    const should_set_validation_params = is_multiplier && is_enabled && new_input_value === '';
+
+    const min_value = validation_params[contract_types[0]]?.[type]?.min;
+    const max_value = validation_params[contract_types[0]]?.[type]?.max;
+    // Storing data from validation params (proposal) in state in case if we got a validation error from BE and proposal stop streaming
+    const [info, setInfo] = React.useState<Record<string, string | undefined>>({ min_value, max_value });
+
+    const input_message =
+        info.min_value && info.max_value && is_enabled ? (
+            <Localize
+                i18n_default_text='Acceptable range: {{min_value}} to {{max_value}} {{currency}}'
+                values={{
+                    currency: currency_display_code,
+                    min_value: info.min_value,
+                    max_value: info.max_value,
+                }}
+            />
+        ) : (
+            ''
+        );
+
+    const updateParentRef = ({ field_name, new_value }: { field_name: string; new_value: string | boolean }) => {
+        if (!parent_ref?.current || !field_name) return;
+        parent_ref.current = { ...parent_ref.current, [field_name]: new_value };
+    };
+
+    const onToggleSwitch = (new_value: boolean) => {
+        is_be_response_received_ref.current = false;
+        setIsEnabled(new_value);
+        updateParentRef({ field_name: is_take_profit_input ? 'has_take_profit' : 'has_stop_loss', new_value });
+
+        if (new_value) {
+            clearTimeout(focus_timeout.current);
+            focus_timeout.current = focusAndOpenKeyboard(focused_input_ref.current, input_ref.current);
+        } else {
+            setErrorText('');
+            updateParentRef({ field_name: is_take_profit_input ? 'tp_error_text' : 'sl_error_text', new_value: '' });
+            input_ref.current?.blur();
+        }
+    };
+
+    // We are using requestPreviewProposal in useEffect in order to validate both fields independently
+    React.useEffect(() => {
```

Comment:
Can we move this to a hook in a separate file? This component is too big with 300+ lines.

Reply:
This file is complicated because of some limitations from BE side. What was done: I've already asked for couple of improvements and additional data in API response such parallel validation of TP and ST (it will help to get rid of requestPreviewProposal and huge useEffect) and min&max available values in error response (it will help to get rid of another useEffect with min_value, max_value).
I can move mentioned useEffect into a hook, but 1) as it's used only in one file it won't be reusable 2) it has a lot of state, setters for state and refs inside, so I need to pass about 10 values into that hook. 
Please, share your thoughts